### DISCUSSION TITLE
Additional memory tests implementation

### DISCUSSION THREAD

**Discussion by x86fr** [memtest86plus/memtest86plus#98] [Enhancement]
The next test that will be implemented is rowhammer (#5). 

Here's a discussion for the other tests that could be implemented in the future.

**Comment by debrouxl** [comment]
SIMD tests. We know that the likes of `movsd`/`movsq` and `stosd`/`stosq` yield sub-par bandwidth on modern CPUs, as you detailed at https://github.com/memtest86plus/memtest86plus/issues/70#issuecomment-1143499381 . Also, this week, I noticed interesting performance numbers for wide-range memory clearing on modern processors in https://lwn.net/Articles/897229/ .

Some usage of SIMD seems to have been envisioned at some point: boot/startup64.S contains a disabled "Enable SSE" section.

**Comment by debrouxl** [comment]
I made a quick & dirty test with 64-bit MMX writes in the 64-bit binary, expecting them not to be of much use - less than 128-bit SSE writes are expected to be, at least.

Diff against the `numa` branch, with superfluous changes to the startup code:
```
diff --git a/boot/startup64.S b/boot/startup64.S
index 2832bef..f43899b 100644
--- a/boot/startup64.S
+++ b/boot/startup64.S
@@ -268,7 +268,7 @@ flush:      movw    $KERNEL_DS, %ax
 
        finit
 
-#if 0
+#if 1
        # Enable SSE.
 
        movq    %cr0, %rax
@@ -276,7 +276,7 @@ flush:      movw    $KERNEL_DS, %ax
        orb     $0x02, %al              # set coprocessor monitoring bit
        mov     %rax, %cr0
        movq    %cr4, %rax
-       orw     $0x0600, %ax            # set OSFXSR and OSXMMEXCPT
+       orb     $0x06, %ah              # set OSFXSR and OSXMMEXCPT
        movq    %rax, %cr4
 #endif
 
diff --git a/system/memrw64.h b/system/memrw64.h
index 3019aa5..7d59772 100644
--- a/system/memrw64.h
+++ b/system/memrw64.h
@@ -14,6 +14,8 @@
 
 #include <stdint.h>
 
+#include <mmintrin.h>
+
 /**
  * Reads and returns the value stored in the 64-bit memory location pointed
  * to by ptr.
@@ -58,6 +60,22 @@ static inline void write64_nt(const volatile uint64_t *ptr, uint64_t val)
     );
 }
 
+/**
+ * Writes val to the 64-bit memory location pointed to by ptr, using MMX register as source operand.
+ */
+#pragma GCC target ("mmx")
+static inline void write64_mm(const volatile uint64_t *ptr, uint64_t val)
+{
+    register __m64 val2 __asm__("%mm0") = (__m64)val;
+    __asm__ __volatile__(
+        "movq %1, %0"
+        :
+        : "m" (*ptr),
+          "y" (val2)
+        : "memory"
+    );
+}
+
 /**
  * Writes val to the 64-bit memory location pointed to by ptr. Reads it
  * back (and discards it) to ensure the write is complete.
diff --git a/tests/mov_inv_fixed.c b/tests/mov_inv_fixed.c
index 5dc47c1..82b7800 100644
--- a/tests/mov_inv_fixed.c
+++ b/tests/mov_inv_fixed.c
@@ -23,7 +23,7 @@
 #include "test_funcs.h"
 #include "test_helper.h"
 
-#define HAND_OPTIMISED  1   // Use hand-optimised assembler code for performance.
+//#define HAND_OPTIMISED  1   // Use hand-optimised assembler code for performance.
 
 //------------------------------------------------------------------------------
 // Public Functions
@@ -84,9 +84,11 @@ int test_mov_inv_fixed(int my_cpu, int iterations, testword_t pattern1, testword
             p = pe;
 #endif
 #else
+#pragma GCC unroll 8
             do {
-                write_word(p, pattern1);
+                write_word_mm(p, pattern1);
             } while (p++ < pe); // test before increment in case pointer overflows
+            __asm__ __volatile__ ("emms");
 #endif
             do_tick(my_cpu);
             BAILOUT;
diff --git a/tests/test_helper.h b/tests/test_helper.h
index 5d38dd6..2562b77 100644
--- a/tests/test_helper.h
+++ b/tests/test_helper.h
@@ -24,6 +24,7 @@
 #define read_word      read64
 #define write_word     write64
 #define write_word_nt  write64_nt
+#define write_word_mm  write64_mm
 #else
 #include "memrw32.h"
 #define read_word      read32
```

I checked that the binary contains the expected `movq mm,...` instructions. For test 3 + 25% of test 4, in NUMA-aware mode:
* on the 4S Opteron 6272 RS904A with 16 x 8 GB PC3-12800R, the difference between `rep stosq`, MMX `movq` in a loop and general purpose `movq` in a loop (keeping `write_word` instead of `write_word_mm`in `mov_inv_fixed.c`), is less than a second over 6'21", so buried within measurement noise;
* on the 4S Xeon E5-4627v2 PowerEdge R820 with 16 x 16 GB PC3-14900R, both MMX `movq` in a loop and general purpose `movq` in a loop are a bit less than 20% slower than `rep stosq`: test 3 executed in 5'05" instead of 4'15". That's not unexpected on ERMS-capable processors.

Next time, I should test MMX `movntq` instead of MMX `movq`, adding a `mfence` alongside the `emms`; however, on these computers, prior tests with general-purpose `movnti` usually showed slowdowns.
Some 32-bit architectures are known to benefit from using MMX `movntq` in `memcpy` implementations (  https://joryanick.com/retro-fast-x86-memcpy.php ). 32-bit architectures aren't primarily the ones for which we should be trying to make memory testing fast, though :)

**Comment by debrouxl** [comment]
And something's wrong with SSE enablement, indeed: all I can get is #UD when executing a `vmovq`. **EDIT: eh, `vmovq` is AVX, not SSE... stupid me.**
Apart from the narrower register operations, the SSE enablement code matches the one at https://wiki.osdev.org/SSE , and tracing CR0 & CR4 shows that the 4 relevant bits are in the appropriate state.
I tried to reinit MXCSR, to no avail.

![memtest86plus_sse_ud_20230928](https://github.com/memtest86plus/memtest86plus/assets/8375/174595d9-7d98-4c27-b5fb-d6259fa4b840)

```
diff --git a/app/interrupt.c b/app/interrupt.c
index 1a337c0..93e1c17 100644
--- a/app/interrupt.c
+++ b/app/interrupt.c
@@ -40,6 +40,11 @@
 #define INT_STKSEGFLT  12
 #define INT_GPF        13
 #define INT_PAGEFLT    14
+#define INT_RESERVED1  15
+#define INT_X87FPE     16
+#define INT_ALIGNCHK   17
+#define INT_MCHCHK     18
+#define INT_SIMDFPE    19
 
 #define OPCODE_HLT      0xF4
 #define OPCODE_JE       0x74
@@ -123,6 +128,9 @@ void interrupt(struct trap_regs *trap_regs)
 {
     // Get the page fault address.
     uintptr_t address = 0;
+    uintptr_t cr0;
+    uintptr_t cr3;
+    uintptr_t cr4;
     if (trap_regs->vect == INT_PAGEFLT) {
 #ifdef __x86_64__
         __asm__(
@@ -136,6 +144,17 @@ void interrupt(struct trap_regs *trap_regs)
         );
 #endif
     }
+#ifdef __x86_64__
+        __asm__(
+            "movq %%cr0, %0; movq %%cr3, %1; movq %%cr4, %2"
+            : "=r" (cr0), "=r" (cr3), "=r" (cr4)
+        );
+#else
+        __asm__(
+            "movl %%cr0, %0; movl %%cr3, %1; movl %%cr4, %2"
+            :"=r" (cr0), "=r" (cr3), "=r" (cr4)
+        );
+#endif
 
     if (trap_regs->vect == INT_NMI) {
         uint8_t *pc = (uint8_t *)trap_regs->ip;
@@ -221,6 +240,10 @@ void interrupt(struct trap_regs *trap_regs)
     display_pinned_message(8, 25, REG_PREFIX "bp: %0" REG_DIGITS "x", (uintptr_t)trap_regs->bp);
     display_pinned_message(9, 25, REG_PREFIX "sp: %0" REG_DIGITS "x", (uintptr_t)trap_regs->sp);
 
+    display_pinned_message(10, 0, " CR0: %0" REG_DIGITS "x", cr0);
+    display_pinned_message(11, 0, " CR3: %0" REG_DIGITS "x", cr3);
+    display_pinned_message(11, 25, "CR4: %0" REG_DIGITS "x", cr4);
+
     display_pinned_message(0, 50, "Stack:");
     for (int i = 0; i < 12; i++) {
         uintptr_t addr = trap_regs->sp + sizeof(reg_t)*(11 - i);
@@ -228,10 +251,10 @@ void interrupt(struct trap_regs *trap_regs)
         display_pinned_message(1 + i, 50, "%0" ADR_DIGITS "x %0" REG_DIGITS "x", addr, (uintptr_t)data);
     }
 
-    display_pinned_message(11, 0, "CS:IP:");
+    display_pinned_message(13, 0, "CS:IP:");
     uint8_t *pp = (uint8_t *)((uintptr_t)trap_regs->ip);
     for (int i = 0; i < 12; i++) {
-        display_pinned_message(11, 7 + 3*i, "%02x", (uintptr_t)pp[i]);
+        display_pinned_message(13, 7 + 3*i, "%02x", (uintptr_t)pp[i]);
     }
 
     clear_screen_region(ROW_FOOTER, 0, ROW_FOOTER, SCREEN_WIDTH - 1);
diff --git a/boot/startup64.S b/boot/startup64.S
index 2832bef..9c6cd4d 100644
--- a/boot/startup64.S
+++ b/boot/startup64.S
@@ -276,8 +276,9 @@ flush:      movw    $KERNEL_DS, %ax
        orb     $0x02, %al              # set coprocessor monitoring bit
        mov     %rax, %cr0
        movq    %cr4, %rax
-       orw     $0x0600, %ax            # set OSFXSR and OSXMMEXCPT
+       orb     $0x06, %ah              # set OSFXSR and OSXMMEXCPT
        movq    %rax, %cr4
+       ldmxcsr init_mxcsr(%rip)
 #endif
 
        # Call the dynamic linker to fix up the addresses in the GOT.
@@ -982,6 +983,9 @@ startup_mutex:
 first_boot:
        .long   1
 
+init_mxcsr:
+       .long   0x1F80
+
        .previous
 
 # Startup stack.
diff --git a/system/memrw64.h b/system/memrw64.h
index 3019aa5..75c6f3d 100644
--- a/system/memrw64.h
+++ b/system/memrw64.h
@@ -14,6 +14,9 @@
 
 #include <stdint.h>
 
+#include <mmintrin.h>
+#include <emmintrin.h>
+
 /**
  * Reads and returns the value stored in the 64-bit memory location pointed
  * to by ptr.
@@ -58,6 +61,38 @@ static inline void write64_nt(const volatile uint64_t *ptr, uint64_t val)
     );
 }
 
+/**
+ * Writes val to the 64-bit memory location pointed to by ptr, using MMX register as source operand.
+ */
+#pragma GCC target ("mmx")
+static inline void write64_mm(const volatile uint64_t *ptr, uint64_t val)
+{
+    register __m64 val2 __asm__("%mm0") = (__m64)val;
+    __asm__ __volatile__(
+        "movq %1, %0"
+        :
+        : "m" (*ptr),
+          "y" (val2)
+        : "memory"
+    );
+}
+
+/**
+ * Writes val to the 128-bit memory location pointed to by ptr, using SSE register as source operand.
+ */
+#pragma GCC target ("sse2")
+static inline void write128_sse(const volatile uint64_t *ptr, uint64_t val)
+{
+    register __m128i val2 __asm__("%xmm0") = _mm_set1_epi64x(val);
+    __asm__ __volatile__(
+        "vmovq %1, %0"
+        :
+        : "m" (*ptr),
+          "x" (val2)
+        : "memory"
+    );
+}
+
 /**
  * Writes val to the 64-bit memory location pointed to by ptr. Reads it
  * back (and discards it) to ensure the write is complete.
diff --git a/tests/mov_inv_fixed.c b/tests/mov_inv_fixed.c
index 5dc47c1..a3dfb21 100644
--- a/tests/mov_inv_fixed.c
+++ b/tests/mov_inv_fixed.c
@@ -23,7 +23,7 @@
 #include "test_funcs.h"
 #include "test_helper.h"
 
-#define HAND_OPTIMISED  1   // Use hand-optimised assembler code for performance.
+//#define HAND_OPTIMISED  1   // Use hand-optimised assembler code for performance.
 
 //------------------------------------------------------------------------------
 // Public Functions
@@ -84,9 +84,11 @@ int test_mov_inv_fixed(int my_cpu, int iterations, testword_t pattern1, testword
             p = pe;
 #endif
 #else
+#pragma GCC unroll 4
             do {
-                write_word(p, pattern1);
+                write_2word_sse(p++, pattern1);
             } while (p++ < pe); // test before increment in case pointer overflows
+            __asm__ __volatile__ ("emms");
 #endif
             do_tick(my_cpu);
             BAILOUT;
diff --git a/tests/test_helper.h b/tests/test_helper.h
index 5d38dd6..33053b0 100644
--- a/tests/test_helper.h
+++ b/tests/test_helper.h
@@ -24,6 +24,8 @@
 #define read_word      read64
 #define write_word     write64
 #define write_word_nt  write64_nt
+#define write_word_mm  write64_mm
+#define write_2word_sse write128_sse
 #else
 #include "memrw32.h"
 #define read_word      read32
```

**Comment by debrouxl** [comment]
The `simd` branch, based on the `numa` branch, contains a prototype implementation of 2 (x86, for now) / 3 (x86-64) new tests: MMX, SSE and AVX temporal + non-temporal variants of test 5. In the startup code, I did not bother detecting the SSE and AVX CPUID bits before enabling these instructions, so as it stands, a binary built from the `simd` branch only works on P4+/Nehalem+/Bulldozer+.

* so far, the only computer where I've seen a speed improvement in tests 11-13 over test 5 is the usual RS904A with 4 x Opteron 6272. The improvement is small, ~1' over ~15', in nontemporal mode; in temporal mode, the improvement is only slightly above measurement noise. MMX, SSE and AVX yield roughly the same execution speed.
* on the R820, which has ERMS, this implementation of SIMD writes is slower than `rep stosq` in temporal mode, and not faster in non-temporal mode. I haven't benchmarked the default NUMA-unaware operation mode, given how slow it is on this platform.
* on a desktop Core i5-2400 (no ERMS) with DDR3-1333, this implementation of SIMD writes isn't faster than `rep stosq` either.
* preceding loops with `#pragma GCC unroll 8/4/2` to write 64 bytes at a time most of the time (I checked the generated code) doesn't change the run time much on any of these computers;
* the 32-bit binary works in QEMU, but I haven't yet benchmarked the code on real 32-bit x86 computers.

```
  tests 5 11 12 13                                               5           11                      12                      13                      5
4S Opteron 6272, 16 x 8 GB PC3-12800R CL11,    numa nontemporal: 00:00:00 -> 00:14:52 (+00:14:52) -> 00:28:30 (+00:13:38) -> 00:42:08 (+00:13:38) -> 00:55:57 (+00:13:49).
4S Opteron 6272, 16 x 8 GB PC3-12800R CL11,    numa            : 00:00:00 -> 00:14:53 (+00:14:53) -> 00:29:35 (+00:14:43) -> 00:44:08 (+00:14:33) -> 00:58:38 (+00:14:30).
4S Opteron 6272, 16 x 8 GB PC3-12800R CL11,    numa nt + unroll: 00:00:00 -> 00:14:54 (+00:14:54) -> 00:28:34 (+00:13:40) -> 00:42:13 (+00:13:39) -> 00:55:52 (+00:13:39).
4S Xeon E5-4627v2, 16 x 16 GB PC3-14900R CL13: numa nontemporal: 00:00:00 -> 00:21:05 (+00:21:05) -> 00:42:11 (+00:21:06) -> 01:03:17 (+00:21:06) -> 01:24:22 (+00:21:05).
4S Xeon E5-4627v2, 16 x 16 GB PC3-14900R CL13: numa            : 00:00:00 -> 00:21:05 (+00:21:05) -> 00:44:26 (+00:23:21) -> 01:07:47 (+00:23:21) -> 01:31:09 (+00:23:22).
```

**Comment by x86fr** [comment]
That looks really promising! I'll give it a try after the first ECC PR.

**Reply by debrouxl** [reply]
Well, I have yet to find a computer where SIMD usage provides a significant speedup :)

On the xSeries 345, 2 x Xeon Gallatin 3.2 GHz (SL72Y, IIRC), in 2S/2C/4T mode:
* temporal MMX, temporal SSE2, non-temporal MMX are just as fast as `rep stosl`;
* temporal SSE2 is slightly faster than `rep stosl`, ~15" over 3'.

On a PIII 1 GHz, the binary barfed a #UD before tests started, on an unprotected `mfence`:
```
00009302 <smp_alloc_mutex>:
    9302:       e8 c5 c3 00 00          call   156cc <__x86.get_pc_thunk.dx>
    9307:       81 c2 3d b6 01 00       add    $0x1b63d,%edx
    930d:       8b 82 3c 17 00 00       mov    0x173c(%edx),%eax
    9313:       85 c0                   test   %eax,%eax
    9315:       8d 48 01                lea    0x1(%eax),%ecx
    9318:       89 8a 3c 17 00 00       mov    %ecx,0x173c(%edx)
    931e:       74 06                   je     9326 <smp_alloc_mutex+0x24>
    9320:       0f ae f0                mfence
    9323:       c6 00 00                movb   $0x0,(%eax)
    9326:       c3                      ret
```
which was consistent with unprotected XMM-using instructions springing in unrelated places:
```
0000b271 <get_xhci_event>:
    b271:       57                      push   %edi
    b272:       89 c1                   mov    %eax,%ecx
    b274:       56                      push   %esi
    b275:       53                      push   %ebx
    b276:       89 d3                   mov    %edx,%ebx
    b278:       8b 90 98 0a 00 00       mov    0xa98(%eax),%edx
    b27e:       89 d0                   mov    %edx,%eax
    b280:       83 e2 0f                and    $0xf,%edx
    b283:       c1 e8 04                shr    $0x4,%eax
    b286:       89 d6                   mov    %edx,%esi
    b288:       c1 e6 04                shl    $0x4,%esi
    b28b:       01 ce                   add    %ecx,%esi
    b28d:       f3 0f 7e 86 c0 02 00    movq   0x2c0(%esi),%xmm0
    b294:       00 
    b295:       66 0f d6 03             movq   %xmm0,(%ebx)
    ...
```
I did not detect that earlier with QEMU, possibly because KVM acceleration was temporarily enabled.

I'm not going to test again on the PIII at this time of the night, but I have pushed the fix for the above issues, as well as the severe build slowdown induced by `#include <immintrin.h>`. I put the include and the SIMD functions inside a `#ifdef` whose corresponding `#define` is present only in `mov_inv_fixed.c` at the moment. I also added `-mno-mmx -mno-sse -mno-sse2` to the CFLAGS in `build32/Makefile`, though that alone does not solve the problem in any way.

It would be great to be able to just use `#include <avxintrin.h>`, as can be done for both `mmintrin.h` and `emmintrin.h` (which includes `mmintrin.h` via `xmmintrin.h`). However, in current GCC, `avxintrin.h` starts with
```
#ifndef _IMMINTRIN_H_INCLUDED
# error "Never use <avxintrin.h> directly; include <immintrin.h> instead."
#endif
```
and `immintrin.h` includes dozens of intrinsics headers, which amount to a whole lot of code :'(
Sure, I could have added `#define _IMMINTRIN_H_INCLUDED` and `#include`s for prerequisite headers before `#include <avxintrin.h>`, but suchlike cheating with system / compiler headers is bad, nasty form because it can break the build at an unspecified time in the future...

**Reply by debrouxl** [reply]
On the VC820 with PIII 1 GHz and 2 x 128 MB RDRAM, non-unrolled non-temporal MMX writes are 8-10% faster than `rep stosl`: 11"-13" over 2'14"-2'15" (for passes beyond 1). Still a pretty small speedup...

Later, I'll check:
* non-unrolled temporal MMX writes;
* 8x unrolled temporal and non-temporal MMX writes;
* in memrw32.h, the usage of SSE `movaps` and `movntps` instead of SSE2 `movntdq` and `movdqa`, in order to unlock test 12 for the PIII and the Athlon XP generations. That is, if I can get the compiler to produce smart code for somehow converting uint32_t into 4 x float stored in a XMM register...

**Reply by debrouxl** [reply]
I made `movaps` and `movntps` work; maybe something slightly better can be done for the u32 -> 4 x f32 conversion.
* on the VC820 with PIII 1 GHz, the non-temporal non-unrolled SSE code path doesn't produce a significant speed difference from the non-temporal non-unrolled MMX code path.
* on the K7S5A with Athlon XP 2400+, DDR-266: well, SSE isn't exposed, though Linux displays the flag in cpuinfo; it looks like it's a known bug: https://www.emutalk.net/threads/your-cpu-doesnt-support-sse-on-athlon-xp-2400.20814/ .
* on the KT4AV with Athlon XP 3000+, DDR-333: non-temporal, non-unrolled MMX and SSE are equally fast, and ~9% faster than `rep stosl`: 17" over 3'05" on the first pass.

**Reply by x86fr** [reply]
The most important metric is not the time needed to complete a test, but the minimum time required to find the first error. Also, SIMD units behavior on modern hardware is vastly different than on very old architecture. Do you have access to some partly defective memory modules, with a couple of wrong bits under stress? I can ask the betatesting team to find you one if needed.

**Reply by martinwhitaker** [reply]
> The most important metric is not the time needed to complete a test, but the minimum time required to find the first error. 

That depends on what you are using it for. In a production environment where you are expecting some failures, that may be true. If you are testing existing machines to check they haven't developed any faults, the time taken for a complete run is important.

**Reply by x86fr** [reply]
Sure, but the time needed for a complete run is mostly arbitrary: We set it to whatever amount of time we think is enough to achieve, let's say, 99% confidence. Any modern computer is able to read/write its entire memory in a few seconds, especially using SIMD where throughput as high as 20 GB/s is common. But my sentence was wrong, the correct one should be : _The most important metric is not the time needed to complete a test, but the minimum time required to achieve 99% confidence that the memory is good or bad._ 

Speaking about that, I noticed that almost every SI factories I'm working with disable test 1/2 because they dramatically increase the time needed to find the first issue (especially on modern computers with many cores). I can't remember a single case where the address tests failed, and I can't think of a case where the computer could boot with a core not being able to access some part of the memory. Maybe we should change the order of the tests to move the address tests after the bit-fade tests. Or only enable the address tests after the first, shortened, pass. What is your opinion on this?

**Reply by debrouxl** [reply]
Usually, the broken UDIMMs I find are of the "computer emits beeps" level of brokenness. My messages indicate that I bought a HMT31GR7CFR4C-PB 8 GB DDR3 RDIMM which had multi-bit uncorrectable errors pinpointed by the boot-time memory tests and subsequently built-in RAM tester ("MpMemory") of the R815 or R820, but I don't remember where that module is, or even whether I kept it... it was in January 2022, in the memtest86+ reboot time frame, and saving defective memory modules for testing purposes wasn't yet natural :)
In any case, when the SIMD code works - it's not there yet - I'll need help for beta-testing. I have low coverage for recent computers: although I have access to Skylake, Coffee Lake or Zen 3, the most recent computer that I can reboot at will is basically the R820, i.e. from a generation nearly a decade old at this point...

Test 1 is already disabled by default at that point, so indeed, why not move the address tests at the end of a pass, after higher-bandwidth memory tests, and enable test 2 only after the first pass.
I've found that using non-temporal SISD writes for the address tests speeds them up (as usual, wall clock time, not time to first error since I can't measure it) on most computers, e.g. shaving dozens of minutes on a 24+h test on 4S Bulldozer, as I had reported. The outlier is the R820, with its strong NUMA effects, unmitigated by the NUMA awareness code, since in that test, all cores are made to access all memory...

I've been working on implementing x86 SSE SIMD reads for the second loop in `test_mov_inv_fixed`. The code works properly for some memory ranges; for other memory ranges, the code finds errors at every address because the value is the logical NOT of the expected value. That would indicate a pattern1 / pattern2 mishap, but I did not forget to write back the inverted value in the code - and if I had, I should be getting errors on every memory range... I'll resume investigation later.
Making the same change in the third loop, which reads and writes memory downwards, should be easy once I make the second loop work. However, ISTR that prefetch engines are not optimized for downwards accesses, so if that also holds for SIMD engines and memory controllers, it may end up being a bad idea for wall clock time...

**Comment by debrouxl** [comment]
After fixing bugs, I've pushed an initial implementation of 128-bit (x86, x86-64) and 256-bit (x86-64) SIMD reads in test_mov_inv_fixed, on top of the 128-bit and 256-bit SIMD writes. On the RS904A, with `numa nontemporal`, test 12 with SSE now runs in ~13'10" and test 13 with AVX in ~13'20", which represents another small memory bandwidth improvement over test 5 running in ~14'55".
I'll test the x86 SSE version later.
EDIT: on the KT4AV with Athlon XP 3000+ and 3 x 512 MB DDR-333: test 5 took 3'34" at pass 1 and 10'43" at pass 2, test 11 nontemporal took 3'14" and 9'42", test 12 nontemporal took 2'25" and 7'30". Still a bit lackluster, but closer to the wall clock time reduction I was hoping that SIMD usage could bring on some microarchitectures :)
EDIT 2: on the VC820 with PIII 1 GHz and 2 x 128 MB RDRAM 700: test 5 took 1'06" at pass 1 and 3'19" at pass 2, test 11 nontemporal took 1'02" and 3'06", test 12 nontemporal took 44" and 2'12". Again, SSE for reads and comparisons shows its benefits here. Too bad it's on a 20+-year-old microarchitecture...
EDIT 3: on a Northwood P4 2.6 GHz with 2 x 512 MB DDR-333: test 5 took 1'35" and 4'45", test 11 nontemporal took 1'20" and 4'01", test 12 nontemporal took 1'21" and 4'03". SSE isn't great for reads here; that processor supports SSE2, though, which means I should probably make 4 SIMD tests instead of 3 SIMD tests...

The build of the SIMD version keeps failing on 64-bit Ubuntu 20.04, which looks very much like https://gcc.gnu.org/bugzilla/show_bug.cgi?id=97148 , fixed in GCC 11+. Ubuntu 20.04 is ancient by now, though... 3 years and a half, plus the pre-release freeze, is a longer interval than the security support period provided by the vast majority of upstream projects.

EDIT: and BTW, while we're at reordering passes... would it make sense to move randomized tests earlier, especially the current test 5 before the current test 4, or even test 8 before the relatively slow test 6 ?

**Reply by debrouxl** [reply]
The SSE and SSE2 tests are now separate.

An amusing test result: the 2S PII Xeon 350 (QS) crashes its own dumb self hard when running the non-temporal version of MMX tests, i.e. `movntq`. AFAICT, it does not even reach the unexpected interrupt handler to display: it simply reboots. MMX `movq` is fine.

It was news to me, but the Internet knows about at least another occurrence of `movntq` crashing an application on PII-based computers, and a possible fix being to replace occurrences of `movntq` by `movq`: https://community.adobe.com/t5/flash-player-discussions/flash-9-0-115-0-crash-with-pentium-ii/m-p/977590 .

**Comment by debrouxl** [comment]
Now that the NUMA awareness code is integrated, I'd like to set my sights on nontemporal and SIMD accesses :)
I've rebased the `simd` branch, and squashed the 4 SIMD-related commits together, in order to make it easier to read the changes. I've performed a tiny bit of cleanup as well.

Separate tests for every alternate way to access memory make sense for validation and benchmarking, but probably less so for production use... This brings questions and options:
* should we really give users the ability to select the alternate instruction types to be tested ?
* should we perform the memory test using all alternate instructions supported by the target platform ? If not, we could perform dynamic benchmarking on memory ranges much larger than caches, i.e. gigabytes on Milan-X and Genoa-X, in order to select the alternate method which maximizes memory bandwidth. Otherwise, doubling the number of iterations for test 5 to 60, which is a multiple of all integers between 2 and 6, and dividing these iterations into 1-6 chunks, one per every supported way to access memory (currently rep movs, MMX, SSE, SSE2, AVX; there's spare room for another one, e.g. AVX512 or MOVDIR, both of which have poor availability for now), may do the job.

On the BDX in my DL380 Gen9, in NUMA-aware mode (of course), all tests 11-14 are slower than test 5's `rep movsq`. Nontemporal mode narrows the gap to 1-2%. I'd expect most ERMS-capable platforms to behave in a similar way ?

**Comment by debrouxl** [comment]
Sub-thread for further discussing reordering tests: we started above but did not reach a conclusion.
In order not to change test numbers, maybe we could run them out of order ? It could prove a support chore, though ("why are tests not executing in order ??")...

**Comment by x86fr** [comment]
To me, the easier PoC to start with is to add 2 variations of Block move, one compiled with AVX2, then other with SSE2 as fallback if CPU doesn't support AVX2. The most modern version of the test (AVX2/SSE2/i586) should be selected automatically at startup. 

It's important to keep in mind that running tests as fast as possible is useless if we start skipping errors. The only important metric is to keep the error detection as high as possible. 

The most important point is to check at various steps that your code offers a better error detection rate. You must have at least 2/3 real-world platforms able to generate memory errors to check that the SSE2/AVX2 code path can offer a better detection rate (number of errors detected by period of time). Overclocking and/or Undervolting is the easiest way to generate memory errors. I'll be happy to help if you need some flawed hardware.

**Reply by debrouxl** [reply]
Block move, eh ? I'll see what I can do.

On the one hand, I do agree that it's certainly high time that I gained access to some flawed hardware. On the other hand... I can't really imagine myself storing 3+ additional platforms around here, so I'd have to shed (boo) or trade some of the ones I already have ^^
Most of my platforms have a peculiarity arguably useful for memtest86+, even if some of them weren't chosen specifically for that purpose: the 2S PII Xeon QS is an early SMP platform where Linux requires `noacpi` to use SMP and nontemporal MMX instructions trigger a reboot, the Optiplex GX1 has a real PIIX4, the VC820 has RDRAM, the xw6600 has DDR2 FB-DIMM, the 2S x 1C Opteron 270 is an early NUMA platform, the desktop K8 has crazy temperature reporting, the UHCI controller in the KT4AV was among the ones which needed USB improvements in the past. I have 3 P4s, two of which are Northwood and have the same UHCI controller (which appeared in this bug tracker alongside the KT4AV) and the third one is Prescott-2M; removing one of the Northwoods should be possible without significantly reducing coverage.

A way not to reduce detection reliability may be to always run SIMD tests, when supported by the target platform, *in addition to* the `rep stos` and `rep movs` variants of `test_mov_inv_fixed` (done) and `test_block_move` (TBD), as part of the same tests. Tests 5 and 7, as currently numbered, aren't that lengthy.
